packages/nimble-components/src/table/specs/table-column-specs/table-column-anchor-hld.md

```diff
@@ -0,0 +1,121 @@
+
+### Angular RouterLink Support
+The real challenge of this column type is integrating with the Angular router. The `RouterLink`/`RouterLinkWithHref` directives are used to intercept clicks on anchors and replace the default navigation action with a call to `Router.navigateByUrl()`. As we have done in the past for other anchor components, we will have our own directive deriving from `RouterLinkWithHref`. Our directive will apply based on the presence of the `nimbleRouterLink` attribute. As we have done in the past, we will also have a directive that throws an error if `routerLink` is used instead. Normally, you would specify the `nimbleRouterLink` and related attributes (e.g. `queryParams`, `replaceUrl`, etc.) directly on the anchor element, but that's not possible for anchors in generated table cells. Our options are to put the directive on `nimble-table` or `nimble-table-column-anchor`. Because we would like to allow different anchor columns to be configured differently, and because it is a more intuitive API, we want to put our directive on the column element:
+```html
+<nimble-table>
+    <nimble-table-column-anchor nimbleRouterLink replaceUrl ...>
+        Link Column 1
+    </nimble-table-column-anchor>
+    ...
+</nimble-table>
+```
+Note that this implies that these configuration parameters must be the same for each link in the column:
+- `target`
+- `queryParams`
+- `queryParamsHandling`
+- `fragment`
+- `state`
+- `relativeTo`
+- `preserveFragment`
+- `skipLocationChange`
+- `replaceUrl`
+
+Normally with RouterLink directives, you assign the url to `routerLink`/`nimbleRouterLink`, but in this case, the url is coming from the table data. `nimbleRouterLink` just has to be present for the directive to be applied.
+
+There is a problem with puting the directive on `nimble-table-column-anchor`, though. `RouterLinkWithHref` works by intercepting a click event that is bubbled up from a decendant link, but the `nimble-table-column-anchor` is not an ancestor of the anchor/table cell. The solution is to introduce a second directive that is on `nimble-table`, which _is_ an ancestor of the anchor/table cell. This directive will be responsible for intercepting click events from anchor elements, then "forwarding" the event to the right column directive.
+
+The first part of this task is to detect which click events have come from a `nimble-anchor`. We can't just look at the event's `target`, because that is changed to the host control upon crossing a shadow DOM boundary. We can, however, look at the `composedPath()` of the event, which will include every element that was along the bubble-up path. If we find a `nimble-anchor` among them (by using `instanceof`), we know we should handle the event.
```

Comment:
> It seems like a bit of an overgeneralization to say that every click event on a `nimble-anchor` will be coming from the hyperlink column.

That's true, but even then, our directive would just fail to find a matching anchor column directive, and it would let the default action occur. That is to say, I don't see any harm in looking at all clicks that came from `nimble-anchor`s.

> Is it worth generalizing our code a bit further and creating a base class that all hyperlink-containing columns could derive from?

I don't see any reason we couldn't do that. But I think we could also just wait until we have another hyperlink-containing column to do it. It shouldn't be any harder to do it then.

Reply:
My main thought is that we don't even know if we'll be the people that need to add that new column. It might be a client team, in which case, they either need to refactor our code to support their use-case, or they are blocked waiting for us to do the refactor.

My opinion is that we should do the extra work now (since I don't think it's really that much extra work) rather than have to change this code at some point in the future.

Reply:
Good point. We can create a base class for the directive.

Reply:
I don't think the table itself should have any specific knowledge of view implementations.

What if we did something like the following:

1. Column plugin authors can say as part of the settings to delegate events with certain names
2. Table listens for events with those names coming from the view instances
3. On the event, the table creates a CustomEvent named `delegated-event` that includes `details: {originalEvent, originalEventName}` and fires it on the column associated with that view
4. The directive can listen for that CustomEvent to access the original event in the details. The directive can manipulate the original event as needed (call preventDefault, etc)


Something like:

```ts
TableColumn {
   abstract readonly columnDelegatedEvents;
}

TextField extends TableColumn {
   readonly columnDelegatedEvents = ['click'] as const;
}
```

Threw the naming together on the fly, mainly trying to discuss the pattern at the moment.

Reply:
Talked with Mert about this proposal:

Clarifications:
- My comment says the table shouldn't have the knowledge of specific views. The PR proposal is that the Angular table directive has knowledge about specific views, not the table.
  - The Angular implementation is still digging into composedPath across the shadow dom boundary. The shadow dom is enacapsulation, we should not break that boundary arbitrarily. Angular bindings should sit in the public client api context not be aware of and coupled to deep internal implementation details of the table such as digging into the shadowroot or accessing internal cell state.

Benefits:
- We can avoid all the internal columnId complexity. I'd rather us make column ids required than jump through those hoops discussed and end up exposing internalUniqueId on the columns anyway.

Concerns:
- The proposed API is Angular specific / we don't need it outside of Angular
  - It's not angular specific, it's a generic way for clients to get the events associated with specific views in a column. I like the mental model, it's like the [event delegation](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_delegation) pattern. Letting authors treat the column element as if it is the parent of it's views in terms of event handling. I think it can be a useful API but I don't have specifics in mind, open to other thoughts or concerns.

Reply:
The only issue I've run into while attempting to prototype this is that our Angular directive needs the `href` of the clicked `nimble-anchor`, and the click event doesn't contain that. It is still possible to use the event's composedPath to get to the originating `nimble-anchor`, but I know we want to avoid that.

Reply:
Discussed with mert offline and the missing piece is that for `details: {originalEvent, originalEventName}` the `originalEvent` should be an event that was captured from the cell-view defined by the column author. That way the `evt.details.originalEvent.target instanceof MyColumnCellView`, i.e. `TableColumnAnchorCellView` in this case.